### ISSUE TITLE
Behavior when applying via plugins block and "apply plugin" are different

### ISSUE THREAD

**Issue by markelliot** [palantir/gradle-processors#12]
When applied through the `plugins` closure on a top level project, the processors jar is not added to Eclipse classpath. In the same project, if applied through `apply plugin`, the processors jar is added to the Eclipse classpath. I expect the latter behavior.

**Comment by cjp39** [comment]
That's really weird. Can you give me more details about your setup — Gradle version, Eclipse version, plugin version?

**Comment by markelliot** [comment]
Tried with:
 - Gradle 2.6 and 2.7
 - Eclipse Mars
 - Processors 1.1

In case it matters, I was trying to use org.immutables:value:2.0.21. 

The failing setup looked like:

    plugins {
        id 'eclipse'
        id 'java'
        id 'org.inferred.processors' version '1.1'
    }

And the working setup like:

    <build script classpath/dep>

    apply plugin: 'eclipse'
    apply plugin: 'java'
    apply plugin: 'org.inferred.processors'

**Comment by cjp39** [comment]
Could you post the output of

    cat gradle/wrapper/gradle-wrapper.properties

**Comment by markelliot** [comment]
distributionBase=GRADLE_USER_HOME
    distributionPath=wrapper/dists
    zipStoreBase=GRADLE_USER_HOME
    zipStorePath=wrapper/dists
    distributionUrl=https\://services.gradle.org/distributions/gradle-2.7-bin.zip

**Comment by markelliot** [comment]
Ok, I think I found it; apparently the order in `plugins` matters:

fails to fix classpath:

    plugins {
        id 'eclipse'
        id 'java'
        id 'org.inferred.processors' version '1.1'
    }

correct classpath:

    plugins {
        id 'java'
        id 'eclipse'
        id 'org.inferred.processors' version '1.1'
    }

**Comment by cjp39** [comment]
O.o omg

**Comment by markelliot** [comment]
Yeah, worse yet, if I mutate the working version to swap eclipse for baseline, things fail again (baseline-eclipse applies the eclipse plugin):

    plugins {
        id 'java'
        id 'com.palantir.baseline-eclipse' version '0.1.1'
        id 'org.inferred.processors' version '1.1'
    }

...but I'll follow up with @uschi2000 about that.

Anyway, short of filing a Gradle bug I don't think there's things to be done here.

**Comment by cjp39** [comment]
I'll leave this open and check there's nothing going wrong in the plugin. Thanks for figuring this out!

**Comment by punya** [comment]
Booooo Gradle.... I thought the whole point of the "declarative" `plugins` syntax was to obviate such considerations :(

**Comment by cjp39** [comment]
We may be doing something dodgy in the gradle-processors plugin, though.

**Comment by uschi2000** [comment]
In the plugins{} syntax, when is Plugin#apply expected to run? I don't think order independence can work once plugins extend behaviour of others.

**Comment by cjp39** [comment]
We take the approach of installing a task that runs after the java/eclipse/etc plugin runs. It may be there is an ordering issue happening there, I suppose.

**Comment by chrisalice** [comment]
Sorry it's taken me so long to address this. I was bamboozled by being apparently unable to replicate it with `apply plugin`, which is necessary for testing. It turns out the order plugins are applied with the `plugins {}` syntax is non-trivial. The example given was:

 ```gradle
plugins {
  id 'eclipse'
  id 'java'
  id 'org.inferred.processors' version '1.1'
}
```

This appears to be equivalent to:

```gradle
apply plugin: 'org.inferred.processors'
apply plugin: 'eclipse'
apply plugin: 'java'
```

or possibly:

```gradle
apply plugin: 'eclipse'
apply plugin: 'org.inferred.processors'
apply plugin: 'java'
```

Weird.

**Comment by jmcampanini** [comment]
could the ordering issues be avoided if we just do the work in an `project.afterEvaluate` block?

**Comment by chrisalice** [comment]
Sadly, no, we need to do the work _before_ the project is evaluated. I have a PR in review that fixes the issue, though.

**Comment by chrisalice** [comment]
This is fixed in [v1.2.3](https://github.com/palantir/gradle-processors/releases/tag/v1.2.3). Sorry for the delay, everyone!
